Source/Scene/ModelExperimental/PickingPipelineStage.js

```diff
@@ -0,0 +1,123 @@
+import Color from "../../Core/Color.js";
+import ComponentDatatype from "../../Core/ComponentDatatype.js";
+import defined from "../../Core/defined.js";
+import Buffer from "../../Renderer/Buffer.js";
+import BufferUsage from "../../Renderer/BufferUsage.js";
+import ShaderDestination from "../../Renderer/ShaderDestination.js";
+
+/**
+ * The picking pipeline stage is responsible for handling picking of primitives.
+ *
+ * @namespace PickingPipelineStage
+ * @private
+ */
+var PickingPipelineStage = {};
+
+/**
+ * Process a primitive. This modifies the following parts of the render resources:
+ * <ul>
+ *  <li>adds attribute and varying declaration for the pick color vertex attribute in the vertex shader for instanced meshes</li>
+ *  <li>adds declaration for the pick color uniform for non-instanced meshes</li>
+ *  <li>adds defines in the shader for when picking is enabled</li>
+ *  <li>creates the pick ID objects in the context</li>
+ * </ul>
+ * @param {PrimitiveRenderResources} renderResources The render resources for this primitive.
+ * @param {ModelComponents.Primitive} node The primitive.
+ * @param {FrameState} frameState The frame state.
+ */
+PickingPipelineStage.process = function (
+  renderResources,
+  primitive,
+  frameState
+) {
+  var context = frameState.context;
+  var runtimeNode = renderResources.runtimeNode;
+  var shaderBuilder = renderResources.shaderBuilder;
+
+  shaderBuilder.addDefine("ALLOWS_PICKING");
+
+  if (defined(runtimeNode.node.instances)) {
+    // For instanced meshes, a pick color vertex attribute is used.
+    processInstancedPickIds(renderResources, context);
+  } else {
+    // For non-instanced meshes, a pick color uniform is used.
+    var model = renderResources.model;
+    var pickObject = {
+      model: model,
+      node: renderResources.runtimeNode,
+      primitive: renderResources.runtimePrimitive,
+    };
+
+    var pickId = context.createPickId(pickObject);
+    model._resources.push(pickId);
+    shaderBuilder.addUniform(
+      "vec4",
+      "czm_pickColor",
+      ShaderDestination.FRAGMENT
+    );
+
+    var uniformMap = renderResources.uniformMap;
+    uniformMap.czm_pickColor = function () {
+      return pickId.color;
+    };
+
+    renderResources.pickId = "czm_pickColor";
+  }
+};
+
+function processInstancedPickIds(renderResources, context) {
+  var instanceCount = renderResources.instanceCount;
+  var pickIds = new Array(instanceCount);
+  var pickIdsTypedArray = new Uint8Array(instanceCount * 4);
+
+  var modelResources = renderResources.model._resources;
+  for (var i = 0; i < instanceCount; i++) {
+    var pickObject = {
+      model: renderResources.model,
+      node: renderResources.runtimeNode,
+      primitive: renderResources.runtimePrimitive,
+      instance: {
+        id: i,
+      },
```

Comment:
Do we expect to add more fields besides `id`? if so, this is okay. if not, maybe just `instanceId`?

Reply:
We could have added the `modelMatrix`, but unfortunately, that won't be available all the time for each instance - so I suppose `instanceId` will work for now.